Auxiliary devices in SOF

ranj063 opened the pull request:
Now added support for nocodec machine driver, ipc test client and probes client all using the auxiliary bus

plbossart commented on include/sound/sof.h:
```diff
@@ -11,6 +11,7 @@
 #ifndef __INCLUDE_SOUND_SOF_H
```
commit message: to an auxiliary driver?

ranj063 commented on sound/soc/sof/sof-audio.c:
```diff
@@ -490,10 +500,46 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
+#endif
+
+	plat_data->pdev_mach = platform_device_register_data(sdev->dev, drv_name,
+							     PLATFORM_DEVID_NONE, mach, size);
```
for the codec machines, we still need this. We'd never get here if nocodec mode is enabled though.

plbossart commented on sound/soc/sof/nocodec.c:
```diff
@@ -81,31 +82,31 @@ int sof_nocodec_setup(struct device *dev,
+module_auxiliary_driver(sof_nocodec_audio)
 
-MODULE_DESCRIPTION("ASoC sof nocodec");
+MODULE_DESCRIPTION("ASoC SOF nocodec machine drv");
```
machine driver

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -490,10 +500,46 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
+#endif
+
+	plat_data->pdev_mach = platform_device_register_data(sdev->dev, drv_name,
+							     PLATFORM_DEVID_NONE, mach, size);
```
why is this still here? It should be conditional when nocodec is not used?

ranj063 replied:
@plbossart it is conditional based on if (!strcmp(drv_name, "snd_sof_nocodec_drv"))

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -465,7 +466,7 @@ int sof_machine_check(struct snd_sof_dev *sdev)
 		return -ENOMEM;
 
-	mach->drv_name = "sof-nocodec";
+	mach->drv_name = "snd_sof_nocodec_drv";
```
I don't get why you need the driver name?

ranj063 replied:
this is required for the ignore_machine logic in the core.

plbossart replied:
sorry, you lost me here. why do we need the 'ignore-machine' logic for a nocodec card.

ranj063 replied:
@plbossart it doesnt matter what card , we use the ignore_machine check here: https://github.com/thesofproject/linux/blob/41fea6ee2ca94d2e69377f3d102f0bb6874d2511/sound/soc/soc-core.c#L1643 
to set the stream name for the DAI links here:
https://github.com/thesofproject/linux/blob/41fea6ee2ca94d2e69377f3d102f0bb6874d2511/sound/soc/soc-core.c#L1702

If this doesnt happen, we'd fail while loading the DAI during topology loading: https://github.com/thesofproject/linux/blob/41fea6ee2ca94d2e69377f3d102f0bb6874d2511/sound/soc/sof/topology.c#L1390

ranj063 replied:
but thinking more about it, maybe we can changethe nocodec machine driver to add the stream names. I'll have to test this out

ranj063 replied:
fixed now

plbossart reviewed (changes requested):
Not fully convinced by the code, the selection between platform device or nocodec mode seems odd.

lyakh commented on sound/soc/sof/sof-audio.c:
```diff
@@ -490,10 +500,47 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
+		plat_data->nocodec_mach = auxdev;
+		dev_dbg(sdev->dev, "created machine %s\n", dev_name(&auxdev->dev));
+
+		return 0;
```
let's put all this in a function, it's taking disproportionately a lot of space for the functionality, only needed for developers. Something like
```
if (!strcmp(drv_name, "sof-nocodec"))
	return nocodec_machine_register(sdev);
```

plbossart commented on sound/soc/sof/nocodec.c:
```diff
@@ -39,6 +40,8 @@ static int sof_nocodec_bes_setup(struct device *dev,
 			return -ENOMEM;
 
+		links[i].stream_name = links[i].name;
+
```
why do we need to add this specifically when changing to the auxiliary_bus?

ranj063 replied:
unfortunately it is needed specifically for auxiliary bus because the naming convention for aux drv uses the KBUILD_MODNAME prefix with the machine driver name provided. But the same change is valid even with the platform driver and probably the right thing to do. Let me submit a separate patch for this.

plbossart replied:
Right, sound good

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -490,10 +500,47 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
-	plat_data->pdev_mach =
-		platform_device_register_data(sdev->dev, drv_name,
-					      PLATFORM_DEVID_NONE, mach, size);
+	/* register machine driver */
```
I know it's the same way as before but technically we register the machine device here.

ranj063 replied:
ack

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -508,6 +555,16 @@ void sof_machine_unregister(struct snd_sof_dev *sdev, void *pdata)
 	struct snd_sof_pdata *plat_data = pdata;
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
+	if (!IS_ERR_OR_NULL(plat_data->nocodec_mach)) {
```
can it be ERR? I only see an assignment in the success case line 535
plat_data->nocodec_mach = auxdev

plbossart reviewed (changes requested):
minor nit-picks, also agree with @lyakh that we should create a helper and avoid putting all this code in the middle of sof_machine_register().

lyakh commented on drivers/base/auxiliary.c:
```diff
@@ -0,0 +1,223 @@
+{
+	struct device *dev = &auxdev->dev;
+
+	dev->bus = &auxiliary_bus_type;
```
I'd still put this under the two checks

plbossart replied:
agree with @lyakh

dmertman replied:
moved to after checks.

lyakh commented on drivers/base/auxiliary.c:
```diff
@@ -0,0 +1,223 @@
+static const struct auxiliary_device_id *auxiliary_match_id(const struct auxiliary_device_id *id,
+							  const struct auxiliary_device *auxdev)
+{
+	while (id->name[0]) {
```
`for (; id->name[0]; id++)`

plbossart replied:
agree with @lyakh this simplifies the loop.

dmertman replied:
changed to this for loop, took out the two id++ lines below and removed the now unnecessary braces in the if(!p) block.

lyakh commented on drivers/base/auxiliary.c:
```diff
@@ -0,0 +1,223 @@
+		return -EINVAL;
+	}
+
+	ret = dev_set_name(dev, "%s.%s.%d", modname, auxdev->name, auxdev->id);
```
don't we need to check `auxdev->name` and `auxdev->id`?

plbossart replied:
@lyakh audev->name is already checked for in the device_init() function (line 144).

I don't think we can check for id, this is allocated by the parent and we can't really check if it's valid. if it's a duplicate the device_add will fail.

dmertman replied:
also, id is an u32 and not a pointer.  If the parent messes up and it is a duplicate, then device_add failure.

lyakh commented on sound/soc/sof/sof-probes-client.c:
```diff
@@ -0,0 +1,424 @@
+	}
+
+	buf[count] = '\0';
+	ret = strsplit_u32((char **)&buf, ",", tkns, num_tkns);
```
is the type-cast really needed? Isn't it the right type as is?

lyakh commented on sound/soc/sof/sof-probes-client.c:
```diff
@@ -0,0 +1,424 @@
+	buf[count] = '\0';
+	ret = strsplit_u32((char **)&buf, ",", tkns, num_tkns);
+exit:
+	kfree(buf);
```
this seems wrong to me. `strsplit_u32()` will update `buf` to point beyond the allocation area. See comment in the function.

lyakh commented on sound/soc/sof/sof-probes-client.c:
```diff
@@ -0,0 +1,424 @@
+ * @num_tkns:	Returned number of tokens obtained.
+ */
+static int
+strsplit_u32(char **buf, const char *delim, u32 **tkns, size_t *num_tkns)
```
I don't see why `buf` has to be `**`, `char *buf` should suffice and would fix the problem with `free()` below.

lyakh reviewed (changes requested):
Comments to the first patch are optional, but it looks like probes need fixing

ranj063 commented:
SOFCI TEST

lyakh commented on drivers/base/auxiliary.c:
```diff
@@ -0,0 +1,285 @@
+	 * The driver name will be <module_name>.<name>, so we need
+	 * to add "2" bytes, one for "." and another for the NULL-termination.
+	 */
+	name = kmalloc(mlen + dlen + 1, GFP_KERNEL);
```
`+2`? But in any case, how about just `kasprintf(GFP_KERNEL, "%s.%s", name, auxdrv->name);`?

plbossart replied:
that sounds much better indeed, good suggestion to use kasprintf and forget having to compute the length,

dmertman replied:
this has already been changed - will be in next push

fredoh9 commented:
I have tested probe with this PR. Working fine, i was able to follow
https://github.com/thesofproject/sof-docs/blob/master/developer_guides/debugability/probes/index.rst

And also I enabled KASAN and check. KASAN didn't complain.

fredoh9 commented:
For alsabat test failures, this is due to we have extra sound card for probes. We have separate discussion.
```
 0 [nocodec        ]: nocodec - nocodec
                      Google-Helios-rev1
 1 [probes         ]: probes - probes
                      Google-Helios-rev1
```

plbossart commented:
> For alsabat test failures, this is due to we have extra sound card for probes. We have separate discussion.
> 
> ```
>  0 [nocodec        ]: nocodec - nocodec
>                       Google-Helios-rev1
>  1 [probes         ]: probes - probes
>                       Google-Helios-rev1
> ```

Yeah this is related to the same problem as when HDMI is defined as a separate card, we need to use dynamic device detection for alsa-bat.

plbossart commented on drivers/base/auxiliary.c:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	if (!auxdev->name) {
+		pr_err("acillary_device has a NULL name\n");
```
typo: auxiliary_device

dmertman replied:
fixed

plbossart commented on drivers/base/auxiliary.c:
```diff
@@ -0,0 +1,285 @@
+module_init(auxiliary_bus_init);
+module_exit(auxiliary_bus_exit);
+
+MODULE_LICENSE("GPL v2");
```
GPL

dmertman replied:
So remove the v2?

The header specifies " GPL-2.0-only" shouldn't the v2 be there?

plbossart commented on sound/soc/sof/Kconfig:
```diff
@@ -50,6 +50,24 @@ config SND_SOC_SOF_DEBUG_PROBES
+	select AUXILIARY_BUS
+	help
+	  This option is not user-selectable but automagically handled by
+	  'select' statements at a higher level
```
add '.' at the end, this was flagged by Randy and fixed in other patches.

plbossart commented on sound/soc/sof/sof-ipc-test-client.c:
```diff
@@ -0,0 +1,329 @@
+		.probe = sof_ipc_test_probe,
+		.remove = sof_ipc_test_remove,
+		.shutdown = sof_ipc_test_shutdown,
+	},
```
don't we need a name?

In the documentation, we state

.driver = {
				.name = AUX_DRV_NAME("myauxiliarydrv"),
			},

I did the same with SoundWire, not sure why  it's not done here?

ranj063 replied:
this part is the new change @plbossart . With KMOD_BUILDNAME, we've made name optional here. If provided, it will get appended with KMOD_BUILDNAME, if not, it will be KMOD_BUILDNAME. It was Leon's suggestion

plbossart replied:
Thanks @ranj063, I will remove this name as well for the SoundWire stuff since I don't need it either.

plbossart commented on sound/soc/sof/intel/Kconfig:
```diff
@@ -57,6 +60,12 @@ config SND_SOC_SOF_INTEL_COMMON
+	tristate
+	help
+	  This option is not user-selectable but automagically handled by
+	  'select' statements at a higher level
```
add '.' at end of sentence

plbossart commented on sound/soc/sof/sof-client.c:
```diff
@@ -0,0 +1,121 @@
+{
+	struct snd_sof_dev *sdev = cdev->sdev;
+
+	lockdep_assert_held(&sdev->client_mutex);
```
why is this not symmetrical with sof_client_dev_register(), is there a reason why the mutex is taken within the register function but outside for unregister?

ranj063 replied:
@plbossart the list iteration is done elsewhere (in the platform-specific code) for unregister. And we need to hold the mutex while iterating through the list. Maybe I should move that code here. let me fix it

plbossart commented on sound/soc/sof/intel/intel-client.c:
```diff
@@ -0,0 +1,53 @@
+{
+	struct sof_client_dev *cdev, *_cdev;
+
+	mutex_lock(&sdev->client_mutex);
```
see comment in previous patch, not sure why we do this here?

if this is by design you should add a comment in the definition of sof_client_dev_unregister().

plbossart commented on sound/soc/sof/Kconfig:
```diff
@@ -42,13 +42,11 @@ config SND_SOC_SOF_OF
-	  Say Y if you want to enable probes.
-	  If unsure, select "N".
+	  This option is not user-selectable but automagically handled by
+	  'select' statements at a higher level
```
'.'

plbossart commented on sound/soc/sof/sof-probes-client.c:
```diff
@@ -0,0 +1,422 @@
+/* driver name will be set based on KBUILD_MODNAME */
+static struct sof_client_drv sof_probes_test_client_drv = {
+	.auxiliary_drv = {
+		.id_table = sof_probes_auxbus_id_table,
```
driver.name?

plbossart commented on sound/soc/sof/intel/intel-client.c:
```diff
@@ -10,6 +10,21 @@
+	mutex_unlock(&sdev->client_mutex);
+}
+EXPORT_SYMBOL_NS_GPL(intel_unregister_client, SND_SOC_SOF_INTEL_CLIENT);
+
```
this should be squashed in the earlier patch where we initially used 'ipc_test'.

plbossart commented on sound/soc/sof/intel/intel-client.c:
```diff
@@ -32,17 +47,7 @@ EXPORT_SYMBOL_NS_GPL(intel_register_ipc_test_clients, SND_SOC_SOF_INTEL_CLIENT);
-	}
-
-	mutex_unlock(&sdev->client_mutex);
+	intel_unregister_client(sdev, "ipc_test");
```
earlier patch

plbossart commented on sound/soc/sof/intel/intel-client.h:
```diff
@@ -11,6 +11,8 @@
 #ifndef __INTEL_CLIENT_H
 #define __INTEL_CLIENT_H
 
+void intel_unregister_client(struct snd_sof_dev *sdev, const char *name);
```
earlier patch

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -411,3 +441,4 @@ const struct sof_intel_dsp_desc jsl_chip_info = {
 };
 EXPORT_SYMBOL_NS(jsl_chip_info, SND_SOC_SOF_INTEL_HDA_COMMON);
 MODULE_IMPORT_NS(SND_SOC_SOF_INTEL_CLIENT);
+MODULE_IMPORT_NS(SND_SOC_SOF_CLIENT);
```
earlier patch?

plbossart replied:
comment missed in updates? This doesn't seem to belong here?

ranj063 replied:
The earlier patch adds inte-client APIs that are common for all platforms and it is already imported in intel-client.c which will be a different module. Now will the probes client registration in cnl, we need this here as well.

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -233,14 +233,44 @@ void cnl_ipc_dump(struct snd_sof_dev *sdev)
+
+static int hda_register_probes_client(struct snd_sof_dev *sdev)
+{
+	return sof_client_dev_register(sdev, "probes", &sof_probes_client_ida);
```
"sof-probes" ?

Or will the name be something like "hda_intel.probes"?

ranj063 replied:
it will be snd_sof.probes

plbossart reviewed (changes requested):
couple of nitpicks mostly, looks good otherwise.

You should really explain more that taking out the IPC test and probe makes the code less monolithic and easier to maintain. It's not obvious in this PR that you are not adding functionality, for example for probes you should state that creating this auxiliary device saves us the hassle of adding probes in 15+ machine drivers.

Thanks @ranj063

ranj063 commented:
> You should really explain more that taking out the IPC test and probe makes the code less monolithic and easier to maintain. It's not obvious in this PR that you are not adding functionality, for example for probes you should state that creating this auxiliary device saves us the hassle of adding probes in 15+ machine drivers.

I have the last statement about not having to modify existing machine drivers for probes. But should I add the first part to the cover-letter or commit message?

ranj063 commented:
@plbossart Dave is addressing the comments in the aux bus implementation. I've fixed all you comments in the SOF patches. Could you please take a look again?

lyakh commented on sound/soc/sof/sof-client.c:
```diff
@@ -122,4 +123,58 @@ struct dentry *sof_client_get_debugfs_root(struct sof_client_dev *cdev)
+{
+	return cdev->sdev->dev;
+}
+EXPORT_SYMBOL_NS_GPL(sof_client_get_dma_dev, SND_SOC_SOF_CLIENT);
```
why was this moved out of `#ifdef`?

lyakh commented on sound/soc/sof/sof-client.h:
```diff
@@ -52,6 +56,27 @@ int sof_client_ipc_tx_message(struct sof_client_dev *cdev, u32 header, void *msg
 			      size_t msg_bytes, void *reply_data, size_t reply_bytes);
 
 struct dentry *sof_client_get_debugfs_root(struct sof_client_dev *cdev);
+struct device *sof_client_get_dma_dev(struct sof_client_dev *cdev);
```
this could be under the `#ifdef` too?

ranj063 replied:
@lyakh this API will be needed for any client that needs to alloc memory pages for DMA and I didnt want to restrict it to just probes alone.

plbossart replied:
probably worth adding a comment that this is generic enough to be used by other clients, and is required by probes to extract the information with DMA transfers.

plbossart commented:
> > You should really explain more that taking out the IPC test and probe makes the code less monolithic and easier to maintain. It's not obvious in this PR that you are not adding functionality, for example for probes you should state that creating this auxiliary device saves us the hassle of adding probes in 15+ machine drivers.
> 
> I have the last statement about not having to modify existing machine drivers for probes. But should I add the first part to the cover-letter or commit message?

I would do both, and add the number of machine drivers (15+) that we have to touch so that reviewers understand the maintenance savings.

ranj063 commented:
> > > You should really explain more that taking out the IPC test and probe makes the code less monolithic and easier to maintain. It's not obvious in this PR that you are not adding functionality, for example for probes you should state that creating this auxiliary device saves us the hassle of adding probes in 15+ machine drivers.
> > 
> > 
> > I have the last statement about not having to modify existing machine drivers for probes. But should I add the first part to the cover-letter or commit message?
> 
> I would do both, and add the number of machine drivers (15+) that we have to touch so that reviewers understand the maintenance savings.

ack!

ranj063 commented:
SOFCI TEST

plbossart commented on sound/soc/sof/sof-ipc-test-client.c:
```diff
@@ -0,0 +1,329 @@
+	 * is enabled. So, increment the usage count to let the device
+	 * suspend after probe is complete.
+	 */
+	pm_runtime_get_noresume(&auxdev->dev);
```
I know I've asked this before, but what does this accomplish if you only do the pm_runtime_enable() line 276.

Wondering if all you need is to replace pm_runtime_put_autosuspend() by pm_runtime_idle(),

plbossart commented on sound/soc/sof/intel/intel-client.c:
```diff
@@ -0,0 +1,44 @@
+void intel_unregister_ipc_test_clients(struct snd_sof_dev *sdev)
+{
+	sof_client_dev_unregister(sdev, "ipc_test", 0);
+	sof_client_dev_unregister(sdev, "ipc_test", 1);
```
wondering how clean it is to use an ida, and later hard-code '0' and '1' here? That doesn't seem super elegant.

Isn't there some sort of macro that tells you which values are in-use?

ranj063 replied:
I've modifed the client APIs a bit, Pierre. For our usecase, we dont really need ida at all. So I just pass hardcoded IDs as we know the number of devices beforehand anyway

plbossart commented on sound/soc/sof/compress.c:
```diff
@@ -13,6 +13,15 @@
+	.pointer	= sof_probe_compr_pointer,
+};
+EXPORT_SYMBOL(sof_probe_compr_ops);
+
```
can you explain why it's no longer dependent on probe-related Kconfigs?

ranj063 replied:
done

plbossart commented on sound/soc/sof/sof-probes-client.c:
```diff
@@ -0,0 +1,422 @@
+	 * is enabled. So, increment the usage count to let the device
+	 * suspend after probe is complete.
+	 */
+	pm_runtime_get_noresume(&auxdev->dev);
```
same comment on pm_runtime, I don't see this as necessary....

plbossart commented on sound/soc/sof/sof-probes-client.c:
```diff
@@ -0,0 +1,422 @@
+	pm_runtime_set_active(&auxdev->dev);
+	pm_runtime_enable(&auxdev->dev);
+	pm_runtime_mark_last_busy(&auxdev->dev);
+	pm_runtime_put_autosuspend(&auxdev->dev);
```
pm_runtime_idle() instead of get_noresume/put_autosuspend?

ranj063 replied:
@plbossart you're right.pm_runtime_idle() does help as well. Fixed now.

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -233,14 +233,44 @@ void cnl_ipc_dump(struct snd_sof_dev *sdev)
+
+static void hda_unregister_probes_client(struct snd_sof_dev *sdev)
+{
+	sof_client_dev_unregister(sdev, "probes", 0);
```
do we actually need an IDA for probes? You could simplify since we only support one!

plbossart reviewed (approved):
I didn't see anything broken @ranj063 but there are still improvements that could be done before sending upstream, e.g. pm_runtime, comments and removing unnecessary 'ida'.

Looks quite good though, thanks for the updates.

ranj063 commented:
@plbossart I've fixed all your comments now. Thanks!

plbossart commented on sound/soc/sof/sof-probes-client.c:
```diff
@@ -0,0 +1,414 @@
+	desc = (struct sof_probe_point_desc *)tkns;
+
+	ret = pm_runtime_get_sync(dev);
+	if (ret < 0) {
```
Need to test for -EACCESS

plbossart commented on sound/soc/sof/sof-probes-client.c:
```diff
@@ -0,0 +1,414 @@
+	}
+
+	ret = pm_runtime_get_sync(dev);
+	if (ret < 0) {
```
-EACCESS test needed

ranj063 replied:
fixed both places

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -235,12 +235,26 @@ void cnl_ipc_dump(struct snd_sof_dev *sdev)
 	intel_unregister_ipc_test_clients(sdev);
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_PROBES)
+	sof_client_dev_unregister(sdev, "probes", 0);
```
ok, so now I am lost here. We don't have an IDA here, but we had one at the client level [1]

plbossart commented on sound/soc/sof/sof-client.c:
```diff
@@ -0,0 +1,116 @@
+	struct auxiliary_device *auxdev = to_auxiliary_dev(dev);
+	struct sof_client_dev *cdev = auxiliary_dev_to_sof_client_dev(auxdev);
+
+	ida_simple_remove(cdev->client_ida, auxdev->id);
```
[1] who allocates the IDA?

ranj063 replied:
good ctach. this is a miss. fixed now.

plbossart reviewed (changes requested):
the IDA part is odd? do we still need it now?
Also -EACCESS is not a problem for pm_runtime, sorry I missed it earlier.

plbossart reviewed (changes requested):
The SOF parts LGTM @ranj063 but still a couple of warnings to fix.

````
CHECK: spaces preferred around that '+' (ctx:VxV)
#555: FILE: drivers/base/auxiliary.c:234:
+		dlen = strlen(auxdrv->name+1);
 		                          ^

Checking sound/soc/sof/sof-probes-client.c ...
sound/soc/sof/sof-probes-client.c:137:10: warning: %d in format string (no. 2) requires 'int' but the argument type is 'unsigned int'. [invalidPrintfArgType_sint]
   ret = snprintf(buf + strlen(buf), remaining,
         ^

sound/soc/sof/sof-ipc-test-client.c:44:14: style: Variable 'cur' is assigned a value that is never used. [unreadVariable]
 ktime_t cur = ktime_get();
             ^

Checking sound/soc/sof/control.c ...
sound/soc/sof/control.c:117:82: style:inconclusive: Function 'snd_sof_volume_info' argument 2 names different: declaration 'ucontrol' definition 'uinfo'. [funcArgNamesDifferent]
int snd_sof_volume_info(struct snd_kcontrol *kcontrol, struct snd_ctl_elem_info *uinfo)
                                                                                 ^
sound/soc/sof/sof-audio.h:128:30: note: Function 'snd_sof_volume_info' argument 2 names different: declaration 'ucontrol' definition 'uinfo'.
   struct snd_ctl_elem_info *ucontrol);
                             ^
sound/soc/sof/control.c:117:82: note: Function 'snd_sof_volume_info' argument 2 names different: declaration 'ucontrol' definition 'uinfo'.
int snd_sof_volume_info(struct snd_kcontrol *kcontrol, struct snd_ctl_elem_info *uinfo)
                                                                                 ^
````

ranj063 commented:
> The SOF parts LGTM @ranj063 but still a couple of warnings to fix.
> 
> ```
> CHECK: spaces preferred around that '+' (ctx:VxV)
> #555: FILE: drivers/base/auxiliary.c:234:
> +		dlen = strlen(auxdrv->name+1);
>  		                          ^
> 
> Checking sound/soc/sof/sof-probes-client.c ...
> sound/soc/sof/sof-probes-client.c:137:10: warning: %d in format string (no. 2) requires 'int' but the argument type is 'unsigned int'. [invalidPrintfArgType_sint]
>    ret = snprintf(buf + strlen(buf), remaining,
>          ^
> 
> sound/soc/sof/sof-ipc-test-client.c:44:14: style: Variable 'cur' is assigned a value that is never used. [unreadVariable]
>  ktime_t cur = ktime_get();
>              ^
> 
> Checking sound/soc/sof/control.c ...
> sound/soc/sof/control.c:117:82: style:inconclusive: Function 'snd_sof_volume_info' argument 2 names different: declaration 'ucontrol' definition 'uinfo'. [funcArgNamesDifferent]
> int snd_sof_volume_info(struct snd_kcontrol *kcontrol, struct snd_ctl_elem_info *uinfo)
>                                                                                  ^
> sound/soc/sof/sof-audio.h:128:30: note: Function 'snd_sof_volume_info' argument 2 names different: declaration 'ucontrol' definition 'uinfo'.
>    struct snd_ctl_elem_info *ucontrol);
>                              ^
> sound/soc/sof/control.c:117:82: note: Function 'snd_sof_volume_info' argument 2 names different: declaration 'ucontrol' definition 'uinfo'.
> int snd_sof_volume_info(struct snd_kcontrol *kcontrol, struct snd_ctl_elem_info *uinfo)
>                                                                                  ^
> ```

fixed the probes abd ipc test client warnings now but control changes are not relevant to this PR. Maybe they need a separate PR

plbossart commented:
> fixed the probes abd ipc test client warnings now but control changes are not relevant to this PR. Maybe they need a separate PR

oh man, someone AGAIN changed stuff without running cppcheck. this is painful. Will send a patch on this one.

plbossart commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -235,12 +235,26 @@ void cnl_ipc_dump(struct snd_sof_dev *sdev)
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_PROBES)
+	sof_client_dev_unregister(sdev, "probes", 0);
+#endif
```
can we add the probes for all of APL/CNL/TGL?

I just tried on TGL and nothing. We should add this for all cAVS1.5+ devices. Thanks!

ranj063 replied:
@plbossart unfortunately Dave is having conflicts with TGL on Linus' tree. Thats why I limited the series to just CNL for now

plbossart replied:
APL should not have any conflicts though?

ranj063 replied:
but the probes functionality was not validated with the APL firmware.

plbossart replied:
that makes no sense to me. Let's test it then. This only depends on the HDaudio DMA...

ranj063 replied:
@fredoh9 is testing it as we speak. All my APL devices gave up on me

fredoh9 replied:
@plbossart I was under impression that the Probes feature is supported from CNL onward.
I built with APL FW with the probes and I'm testing on UP2. Everything seems to be working fine, except crecord's output file is not growing. I set multiple extraction points and it ends up with setting all of buffer widget IDs but no good news. I need to check with FW team for some clarification. I will update again.

fredoh9 replied:
Probes is tested in APL UP2. I missed `aplay` when I tested in APL. That's why its output didn't grow.
Below the diff.txt from @plbossart, it looks great.

plbossart commented:
@ranj063 suggested patch: [apl-cnl-tgl.diff.txt](https://github.com/thesofproject/linux/files/5426032/apl-cnl-tgl.diff.txt)

libinyang commented:
@ranj063 I tested this patch on my sdw platform, and can't find the 'sof-probes' card in /proc/asound/cards. Do it correctly or not? PS: I'm using the kernel: https://sof-ci-storage.sh.intel.com/pr/linux/pr2501-4763/

plbossart commented:
> @ranj063 I tested this patch on my sdw platform, and can't find the 'sof-probes' card in /proc/asound/cards. Do it correctly or not? PS: I'm using the kernel: https://sof-ci-storage.sh.intel.com/pr/linux/pr2501-4763/

@libinyang did you test on a CML platform? If no, probes are not added, see patch above.

fredoh9 commented:
@libinyang which platform did you try? If you keep not seeing Probe card, I can help you to figure out why Probe soundcard is not enumerated.

libinyang commented:
@plbossart @fredoh9 I tested it on TGL, not CML. Thanks for the information. I will look for a CML SDW platform to have a test on the userspace, especially for the pulseaudio test.
